Cloud Freemium - Integrations

mickmister opened the pull request:
#### Summary

This PR makes it so the server will stop an admin from enabling a new integration, if the server has reached its limit of enabled integrations. When a plugin is enabled or disabled, the server will notify the client of the new integration count, in order to display the updated count in the Subscriptions page.

These plugins are ignored when counting the number of enabled integrations:

- `playbooks`
- `focalboard`
- `com.mattermost.apps`
- `com.mattermost.nps`
- `com.mattermost.plugin-channel-export`
- `com.mattermost.calls`

#### Ticket Link

Epic - https://mattermost.atlassian.net/browse/MM-43851

https://mattermost.atlassian.net/browse/MM-43852
https://mattermost.atlassian.net/browse/MM-43854
https://mattermost.atlassian.net/browse/MM-43857
https://mattermost.atlassian.net/browse/MM-43859
https://mattermost.atlassian.net/browse/MM-43861

#### Related Pull Requests

- https://github.com/mattermost/mattermost-webapp/pull/10301

#### Release Note

```release-note
A server configured to use the Cloud Freemium limits now guards against enabling more integrations than the configured limit.
```

mickmister commented on api4/cloud.go:
```diff
@@ -41,6 +41,9 @@ func (api *API) InitCloud() {
 	api.BaseRoutes.Cloud.Handle("/subscription", api.APISessionRequired(changeSubscription)).Methods("PUT")
 
+	// GET /api/v4/cloud/integrations
+	api.BaseRoutes.Cloud.Handle("/installed_integrations", api.APISessionRequired(getInstalledIntegrations)).Methods("GET")
```
@neallred `/api/v4/cloud` may not be the right spot for this (also I see I have a typo in the comment). At the moment, is no general purpose subpath for querying integrations. They are currently segmented between `api/v4/plugins` and `/plugins/com.mattermost.apps`. I'm not sure if this endpoint warrants creating a `api4/integrations.go` file.

marianunez replied:
> , but at the same time this is a request specific to cloud instances of Mattermost, so it makes sense here, too.

It seems `getInstalledIntegrations` doesn't seem to have anything specific to cloud or CWS requests. 2/5 that it would make sense more under a new `/limits` subpath or even under a more integrations related endpoint. This information will still be useful in other scenarios outside of cloud like for example an Insights type of dashboard in System Console

neallred commented on app/integrations.go:
```diff
@@ -0,0 +1,170 @@
+			enabledApps = []ListedApp{}
+		}
+
+		for _, ap := range enabledApps {
```
Can't use the same set trick down here (unless we build a set of the currently added integrations, but not sure I like that). But we could use a [labeled loop](https://www.ardanlabs.com/blog/2013/11/label-breaks-in-go.html).

```go
	AppsLoop:
		for _, ap := range enabledApps {
			for _, integration := range out {
				if integration.ID == ap.Manifest.AppID {
					integration.Type = "plugin-app"
					continue AppsLoop
				}
			}

			integration := &model.InstalledIntegration{
				Type:    "app",
				ID:      ap.Manifest.AppID,
				Name:    ap.Manifest.DisplayName,
				Version: ap.Manifest.Version,
				Enabled: ap.Enabled,
			}

			out = append(out, integration)
		}
```

neallred commented on app/integrations.go:
```diff
@@ -0,0 +1,170 @@
+			ignore := false
+			for _, integration := range out {
+				if integration.ID == ap.Manifest.AppID {
+					integration.Type = "plugin-app"
```
For my own curiosity and understanding, what are examples of a plugin and an app both having the same id? Are there some plugins that rely on having a matching app, and so they should only be counted as a single integration?

neallred commented on app/plugin.go:
```diff
@@ -457,6 +476,20 @@ func (ch *Channels) disablePlugin(id string) *model.AppError {
+
+	message := model.NewWebSocketEvent(model.WebsocketEventInstalledIntegrationsChanged, "", "", "", nil)
+	message.Add("integrations", integrations)
+	message.GetBroadcast().ContainsSensitiveData = true
```
Ditto comment about needing an (updated) enabled plugin count for regular users. I'd be ok modifying this to only return the count or to have two events, one that returns this and one that returns the count of enabled.

neallred commented on app/integrations.go:
```diff
@@ -0,0 +1,173 @@
+		return model.NewAppError("checkIfIntegrationMeetsFreemiumLimits", "api.cloud.request_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	if limits == nil || limits.Integrations == nil || limits.Integrations.Enabled == nil {
```
:100:

neallred commented on app/plugin.go:
```diff
@@ -370,7 +374,17 @@ func (a *App) GetActivePluginManifests() ([]*model.Manifest, *model.AppError) {
 // Notifies cluster peers through config change.
 func (a *App) EnablePlugin(id string) *model.AppError {
-	return a.ch.enablePlugin(id)
+	appErr := a.checkIfIntegrationMeetsFreemiumLimits(id)
```
Should app/onboarding.go also perform this check, and/or should it be updated to use this public EnabledPlugin method? I'm imagining it would be especially relevant if/when the onboarding screen brings up a user-selectable menu of many plugins; though possibly even now a tricky user could create an abusive payload with too many plugins to be enabled for onboarding.

mickmister replied:
@neallred I think calling the public method makes the most sense. This scenario means the admin went out of their way to sabotage the process, so it's appropriate to fail. Ideally it shouldn't affect other parts of onboarding.

neallred reviewed (changes requested):
In general looks good! The biggest thing from a consuming side is regular users needing some way to get the count.

mickmister commented on app/integrations.go:
```diff
@@ -0,0 +1,170 @@
+	}
+
+	for _, p := range plugins {
+		ignore := false
```
@neallred Although the nested loop is a little ugly, that's how most of the codebase is written, and is common with Go code AFAIK. 0/5 on whether we should have it either way. Do we have a set pattern for this in our codebase? (No pun intended)

neallred replied:
I see what you did there :smile: The set pattern is also common in go, and we use it in a handful of places in mattermost-server:

```
app/channel.go
3185:	channelMemberMap := make(map[string]struct{})

app/product.go
32:	pmap := make(map[string]struct{})

store/sqlstore/channel_store.go
1458:		visited := make(map[string]struct{})

services/sharedchannel/sync_send.go
79:	notified := make(map[string]struct{})

model/user.go
853:var restrictedUsernames = map[string]struct{}{
```

The empty struct doesn't take space, so this set structure should be comparable in size to a slice.

Its good to avoid nested loops for brevity, readability, and because they are often quadratic time. For example, every time we add a new system plugin, it is likely to exist in the plugins we fetch, and will always exist in the list/set of system plugins. So if we have 5 system plugins, that's up to 25 iterations between the two loops. If we increase to 6, that's up to 36. If we have the set, it becomes linear, 5 -> 6. In this case, the number of plugins is small enough that it shouldn't matter much for performance either way.

hanzei commented on app/plugin.go:
```diff
@@ -370,7 +374,17 @@ func (a *App) GetActivePluginManifests() ([]*model.Manifest, *model.AppError) {
 func (a *App) EnablePlugin(id string) *model.AppError {
-	return a.ch.enablePlugin(id)
+	appErr := a.checkIfIntegrationMeetsFreemiumLimits(id)
+	if appErr != nil {
+		return appErr
+	}
```
Should `checkIfIntegrationMeetsFreemiumLimits` also be checked when a plugin get installed? How does the UX look like if an admin installs a plugin via the Marketplace that would push the server over the limit?

catalintomai replied:
@mickmister - is the current design that we notify if new plugins/apps get installed and we are over the limit? I thought that we only do that when we enable installed plugins - user can install as many as they please.

hanzei commented on app/integrations.go:
```diff
@@ -0,0 +1,168 @@
+}
+
+func (ch *Channels) getInstalledApps() ([]ListedApp, *model.AppError) {
+	rawURL := "/plugins/com.mattermost.apps/api/v1/marketplace"
```
Makes sense :+1:

hanzei commented on app/integrations.go:
```diff
@@ -0,0 +1,173 @@
+func (ch *Channels) getInstalledApps() ([]ListedApp, *model.AppError) {
+	rawURL := "/plugins/com.mattermost.apps/api/v1/marketplace"
+	values := url.Values{
+		"include_plugins": []string{"true"},
```
Why are plugin apps included if the list of plugins is also fetched in `getInstalledIntegrations`?

DHaussermann commented:
/update-branch

neallred commented on model/config.go:
```diff
@@ -2786,34 +2786,34 @@ func (s *PluginSettings) SetDefaults(ls LogSettings) {
+	if s.PluginStates[PluginIdNPS] == nil {
 		// Enable the NPS plugin by default if diagnostics are enabled
-		s.PluginStates["com.mattermost.nps"] = &PluginState{Enable: ls.EnableDiagnostics == nil || *ls.EnableDiagnostics}
+		s.PluginStates[PluginIdNPS] = &PluginState{Enable: ls.EnableDiagnostics == nil || *ls.EnableDiagnostics}
```
:+1: to constants refactor

neallred reviewed (approved):
LGTM, thanks @mickmister !

mickmister commented:
@neallred I've added some checks to handle the case where the admin enables a plugin using the `enable` bit of a given plugin's settings. Please let me know if you'd like me to put this in a separate PR.

cc @DHaussermann

neallred commented on app/integrations.go:
```diff
@@ -26,6 +26,21 @@ type ListedApp struct {
+	for pluginId, newState := range newConfig.PluginSettings.PluginStates {
+		oldState, ok := oldConfig.PluginSettings.PluginStates[pluginId]
+		if newState.Enable && !(ok && oldState.Enable) {
+			return a.checkIfIntegrationMeetsFreemiumLimits(pluginId)
```
I think we only want to return here if it is an `error`. Otherwise, if there are multiple plugins that would newly be enabled, we will end up enabling them all even if only the first one should be eligibile for enabling.

neallred replied:
Nevermind, was reading an earlier draft instead of the latest diff. Looks good here now.

neallred replied:
Double nevermind: I think my first comment still stands.

neallred commented:
Adding the checks in this PR sounds good to me. Besides what I think is an error in the loop that calls `a.checkIfIntegrationMeetsFreemiumLimits(pluginId)` and some CI pipeline errors, this looks good to me.

neallred reviewed (approved):
Changes look good to me. Thanks, @mickmister !

mickmister commented:
@neallred Would you happen to know how to solve this CI issue? It looks like the license may be missing, or maybe it's configured incorrectly for the unit tests in the `api4` package:

```
{"timestamp":"2022-05-18 08:42:02.648 Z","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:156","method":"POST","url":"/api/v4/users/login","request_id":"nijf9sypst89tywxh5yw5x66yo","status_code":"200"}
{"timestamp":"2022-05-18 08:42:02.649 Z","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:156","method":"GET","url":"/api/v4/usage/integrations","request_id":"3xtog4hsopngt8xay36krp7xtw","status_code":"200"}
    usage_test.go:62: 
{"timestamp":"2022-05-18 08:42:02.676 Z","level":"info","msg":"Purging all caches","caller":"app/admin.go:172"}
{"timestamp":"2022-05-18 08:42:02.677 Z","level":"debug","msg":"We could not find the license key in the database or on disk at","caller":"utils/license.go:114","filename":"/tmp/testlib3944073384/config/mattermost.mattermost-license"}
{"timestamp":"2022-05-18 08:42:02.677 Z","level":"error","msg":"License key from https://mattermost.com required to unlock enterprise features.","caller":"app/license.go:147","error":"resource: License id: "}
{"timestamp":"2022-05-18 08:42:02.677 Z","level":"info","msg":"Stopping Server...","caller":"app/server.go:972"}
```

neallred commented:
I do see those logs, but I don't think they are the cause of failues. In the latest [test run](https://circleci.com/api/v1.1/project/github/mattermost/mattermost-server/273005/output/104/0?file=true&allocation-id=6287b226837d7b2fb8543ee8-0-build%2FBMYNKG4F), looks like there is an error about mismatched types in expects (int vs int64), and an error about counts.
```
    usage_test.go:62: ^M
          Error Trace:  usage_test.go:62^M
          Error:        Not equal: ^M
                        expected: int(0)^M
                        actual  : int64(0)^M
          Test:         TestGetIntegrationsUsage/good_request_returns_response^M
{"timestamp":"2022-05-20 15:34:27.507 Z","level":"info","msg":"Purging all caches","caller":"app/admin.go:172"}^M
{"timestamp":"2022-05-20 15:34:27.508 Z","level":"debug","msg":"We could not find the license key in the database or on disk at","caller":"utils/license.go:114","filename":"/tmp/testlib3144747731/config/mattermost.mattermost-license"}^M
{"timestamp":"2022-05-20 15:34:27.508 Z","level":"error","msg":"License key from https://mattermost.com required to unlock enterprise features.","caller":"app/license.go:147","error":"resource: License id: "}^M
{"timestamp":"2022-05-20 15:34:27.509 Z","level":"info","msg":"Stopping Server...","caller":"app/server.go:972"}^M
{"timestamp":"2022-05-20 15:34:27.509 Z","level":"info","msg":"stopping websocket hub connections","caller":"app/web_hub.go:113"}^M
{"timestamp":"2022-05-20 15:34:27.510 Z","level":"info","msg":"Shutting down plugins","caller":"app/plugin.go:340"}^M
{"timestamp":"2022-05-20 15:34:27.510 Z","level":"debug","msg":"Disabling plugin health check job","caller":"plugin/environment.go:530"}^M
{"timestamp":"2022-05-20 15:34:27.510 Z","level":"info","msg":"Server stopped","caller":"app/server.go:1061"}^M
--- FAIL: TestGetIntegrationsUsage (0.90s)^M
    --- PASS: TestGetIntegrationsUsage/unauthenticated_users_can_not_access (0.34s)^M
    --- FAIL: TestGetIntegrationsUsage/good_request_returns_response (0.55s)^M
=== RUN   TestCreateUser^M
```

```
{"timestamp":"2022-05-20 11:25:35.061 -05:00","level":"info","msg":"Syncing plugin from file store","caller":"app/plugin.go:320","bundle":"plugins/testplugin.tar.gz"}
    plugin_test.go:1026:
          Error Trace:  plugin_test.go:1026
          Error:        "checkIfIntegrationMeetsFreemiumLimits: You've reached the max limit of 1 installed integrations. Please disable or uninstall another integration to enable this one., " does not contain "Too many enabled integrations"
          Test:         TestEnablePluginWithCloudLimits
{"timestamp":"2022-05-20 11:25:35.173 -05:00","level":"info","msg":"Purging all caches","caller":"app/admin.go:172"}
{"timestamp":"2022-05-20 11:25:35.174 -05:00","level":"debug","msg":"We could not find the license key in the database or on disk at","caller":"utils/license.go:114","filename":"/tmp/testlib3949511218/config/mattermost.mattermost-license"}
{"timestamp":"2022-05-20 11:25:35.175 -05:00","level":"error","msg":"License key from https://mattermost.com required to unlock enterprise features.","caller":"app/license.go:147","error":"resource: License id: "}
{"timestamp":"2022-05-20 11:25:35.176 -05:00","level":"debug","msg":"Logging metrics enabled","caller":"app/server.go:944"}
{"timestamp":"2022-05-20 11:25:35.176 -05:00","level":"info","msg":"Stopping Server...","caller":"app/server.go:972"}
{"timestamp":"2022-05-20 11:25:35.176 -05:00","level":"info","msg":"stopping websocket hub connections","caller":"app/web_hub.go:113"}
{"timestamp":"2022-05-20 11:25:35.178 -05:00","level":"info","msg":"Shutting down plugins","caller":"app/plugin.go:340"}
{"timestamp":"2022-05-20 11:25:35.178 -05:00","level":"debug","msg":"Disabling plugin health check job","caller":"plugin/environment.go:530"}
{"timestamp":"2022-05-20 11:25:35.178 -05:00","level":"info","msg":"Server stopped","caller":"app/server.go:1061"}
--- FAIL: TestEnablePluginWithCloudLimits (0.61s)
FAIL
```

mickmister commented:
Thanks @neallred :+1: Tests are passing now

catalintomai commented on app/plugin.go:
```diff
@@ -156,6 +156,10 @@ func (ch *Channels) syncPluginsActiveState() {
 	}
+
+	if err := ch.notifyIntegrationsUsageChanged(); err != nil {
+		mlog.Warn("failed to notify installed integrations changed", mlog.Err(err))
```
is it `installed` or `activated`?

catalintomai commented on model/client4.go:
```diff
@@ -8110,3 +8110,16 @@ func (c *Client4) GetPostsUsage() (*PostsUsage, *Response, error) {
 	return usage, BuildResponse(r), err
 }
+
+// GetIntegrationsUsage returns usage information on integrations, including the count of enabled integrations
```
not clear, is `including` exclusively enabled integrations or does "usage information" also contains other info being returned?

catalintomai commented on app/plugin_install.go:
```diff
@@ -447,6 +464,10 @@ func (ch *Channels) RemovePlugin(id string) *model.AppError {
 	}
 
+	if err := ch.notifyIntegrationsUsageChanged(); err != nil {
+		mlog.Warn("Failed to notify plugin status changed", mlog.Err(err))
```
is it really "plugin status changed" or "installed integrations changed'?

DHaussermann commented:
/update-branch

mickmister commented:
@neallred I've made some changes to the PR since your last review:

- I changed the `IntegrationsUsage` struct to match the `ProductLimits` struct, so they both have `Enabled int`, rather than the usage struct having `Count int`. This makes it in-line with what the frontend stores in redux. The webapp PR is also updated https://github.com/mattermost/mattermost-webapp/pull/10301
- I removed the check for exceeding the limits in the plugin upload endpoint. We don't allow plugin uploads in Cloud, so this check is unnecessary. 
- I removed the fetching of Apps from the Apps plugin, because the endpoint being used at the moment to query information on installed Apps requires a user's token. This was working initially when we were requiring a user-specific `request.Context`, but it no longer works for the more low-level checks of the integrations usage. Keeping the code in here would result in a useless warning logged every time the limits are checked.
    - As a follow up, I will instead create a usage endpoint on the Apps plugin specific to this use case, and return just the count of enabled Apps. It may not need to be an authenticated endpoint, but if we want that endpoint to be authenticated, we will need to provide a way for the server to communicate with plugins in a trusted way where the plugin knows "this is a trusted request coming from the server, not tied to a specific user". Either way, it is outside of the scope of this PR.

amyblais commented:
> @mickmister @amyblais Should this be cherry-picked to `release-7.0` / `cloud` branches as well? Asking cause I am facing some conflicts with #20388 due to having built my PR on top of this (changes to plugin ids). Just want to avoid introducing even more.

This is already in ``cloud``. @mickmister can confirm how risky this is to cherry-pick to `release-7.0`.

mickmister commented:
/cherry-pick release-7.0

mattermod commented:
Cherry pick is scheduled.

mattermod commented:
Error trying doing the automated Cherry picking. Please do this manually

```
+++ Updating remotes...
Fetching upstream
Failed to add the RSA host key for IP address '140.82.112.4' to the list of known hosts (/app/.ssh/known_hosts).
Fetching origin
Failed to add the RSA host key for IP address '140.82.113.4' to the list of known hosts (/app/.ssh/known_hosts).
+++ Updating remotes done...
+++ Creating local branch automated-cherry-pick-of-mattermost-server-#20185-upstream-release-7.0-1654624938
Switched to a new branch 'automated-cherry-pick-of-mattermost-server-#20185-upstream-release-7.0-1654624938'
Branch 'automated-cherry-pick-of-mattermost-server-#20185-upstream-release-7.0-1654624938' set up to track remote branch 'release-7.0' from 'upstream'.

+++ About to attempt cherry pick of PR #20185 with merge commit ea5bfe3fd95697d8c816bfeae494fc3fab19a1c3.

error: could not apply ea5bfe3fd... Cloud Freemium - Integrations (#20185)
hint: after resolving the conflicts, mark the corrected paths
hint: with 'git add <paths>' or 'git rm <paths>'
hint: and commit the result with 'git commit'

+++ Conflicts detected:

UU model/client4.go
Aborting.

+++ Aborting in-progress git cherry-pick.

+++ Returning you to the master branch and cleaning up.

```

mickmister commented:
@amyblais @streamer45 This should be good to cherry-pick to `release-7.0`. I'm taking care of this now. cc @DHaussermann

mickmister commented:
@neallred In the process of cherry-picking this to `release-7.0` (to avoid merge conflicts with related efforts), I see some files are deleted on that branch during the cherry-pick, including `api4/usage*`, `model/usage.go`, and some code removed in `model/client4.go`. What do you advise for a path forward here?

streamer45 commented:
> @streamer45 @neallred I think it makes sense to _not_ cherry-pick this PR actually, considering the complications mentioned above and the time constraints of code freeze. What do the merge conflicts look like on your end @streamer45? If it's only a few lines, I suggest we leave this PR out and deal with the plugin ID naming conflicts in your cherry-pick.

Well, sooner or later these will need to be addressed I guess. Just wanted to avoid double work for whoever ends up merging. I am good to fix the conflicts on my end, they look trivial.
